style_guides/js_style_guide.md

```diff
@@ -153,188 +185,288 @@ Variables, properties and function names should use `lowerCamelCase`.  They
+const bar = true;
 
-function File() {
-}
-File.fullPermissions = 0777;
-```
+// good
+const foo = {
+  bar
+};
 
-[const]: https://developer.mozilla.org/en/JavaScript/Reference/Statements/const
+// bad
+const foo = {
+  bar: bar
+};
 
-## Magic numbers
+// also bad
+const foo = {
+  baz: false,
+  bar
+};
+```
 
-These are numbers (or other values) simply used in line in your code. **Do not use these**, give them a variable name so they can be understood and changed easily.
+## Modules
 
-*Right:*
+Module dependencies should be written using native ES2015 syntax wherever
+possible (which is almost everywhere):
 
 ```js
-var minWidth = 300;
+// good
+import { mapValues } from 'lodash';
+export default mapValues;
 
-if (width < minWidth) {
+// bad
+const _ = require('lodash');
+module.exports = _.mapValues;
+
+// worse
+define(['lodash'], function (_) {
   ...
-}
+});
 ```
 
-*Wrong:*
+In those extremely rare cases where you're writing server-side JavaScript in a
+file that does not pass run through webpack, then use CommonJS modules.
+
+In those even rarer cases where you're writing client-side code that does not
+run through webpack, then do not use a module loader at all.
+
+## Import only top-level modules
+
+The files inside a module are implementation details of that module. They
+should never be imported directly. Instead, you must only import the top-level
+API that's exported by the module itself.
+
+On the other hand, a module should be able to import parent and sibling
+modules.
 
 ```js
-if (width < 300) {
-  ...
-}
+// good
+import foo from 'foo';
+import child from './child';
+import parent from '../';
+import ancestor from '../../../';
+import sibling from '../foo';
+
+// bad
+import inFoo from 'foo/child';
+import inSibling from '../foo/child';
 ```
 
 ## Global definitions
 
-Don't do this. Everything should be wrapped in a module that can be depended on by other modules. Even things as simple as a single value should be a module.
+Don't do this. Everything should be wrapped in a module that can be depended on
+by other modules. Even things as simple as a single value should be a module.
 
 ## Function definitions
 
-Prefer the use of function declarations over function expressions. Function expressions are allowed, but should usually be avoided.
-
-Also, keep function definitions above other code instead of relying on function hoisting.
+Use function declarations over function expressions.
 
-*Preferred:*
+Also, keep function definitions above other code instead of relying on function
+hoisting.
 
 ```js
+// good
 function myFunc() {
   ...
 }
+
+// bad
+const myFunc = function () {
+  ...
+};
 ```
 
-*Allowed:*
+## Arrow functions
+
+If you must use a function expression, then use an arrow function:
 
 ```js
-var myFunc = function () {
-  ...
-};
+// good
+[1, 2, 3].map((n) => {
+  const m = doSomething(n);
+  return m - n;
+})
+
+// bad
+[1, 2, 3].map(function (n) {
+  const m = doSomething(n);
+  return m - n;
+})
```

Comment:
If a function warrants a name, it should be a function declaration (section above).